src/concurrent/squeue.rs

```diff
@@ -67,18 +64,63 @@ impl SubmissionQueue<'_> {
     /// Developers must ensure that parameters of the [`Entry`] (such as buffer) are valid and will
     /// be valid for the entire duration of the operation, otherwise it may cause memory problems.
     pub unsafe fn push(&self, Entry(entry): Entry) -> Result<(), Entry> {
-        let _lock = self.push_lock.lock();
-
         let head = (*self.queue.head).load(atomic::Ordering::Acquire);
-        let tail = unsync_load(self.queue.tail);
 
-        if tail.wrapping_sub(head) == self.ring_entries {
-            return Err(Entry(entry));
+        let previous_reserved_tail = self
+            .reserved_tail
+            .fetch_update(
+                atomic::Ordering::Acquire,
+                atomic::Ordering::Relaxed,
+                |tail| {
+                    if tail.wrapping_sub(head) == self.ring_entries {
+                        None
+                    } else {
+                        Some(tail.wrapping_add(1))
+                    }
+                },
+            )
+            .map_err(|_| Entry(entry))?;
+
+        *self
+            .queue
+            .sqes
+            .add((previous_reserved_tail & self.ring_mask) as usize) = entry;
+
+        while (*self.queue.tail)
+            .compare_exchange(
+                previous_reserved_tail,
+                previous_reserved_tail.wrapping_add(1),
+                atomic::Ordering::Release,
+                atomic::Ordering::Relaxed,
+            )
+            .is_err()
+        {
+            libc::syscall(
```

Comment:
I think we shouldn't wait when tail == previous_reserved_tail?

Reply:
Yes, my bad - I've fixed that now.